File: desktop/src/main/analytics-service.ts
// Sends one /app/heartbeat per UTC day, gated by an opt-out toggle.
// WHY once per day is a CONTRACT, not a tuning knob: AboutPopup.tsx promises
// users "anonymous usage data … once per day". Sending more often (to fix a
// counting gap, say) rewrites that copy — a product decision. The day boundary
// is UTC on every device; the owner dashboard chooses which clock it counts by.
// Identity is HMAC_SHA256(SALT, machine_id || platform), computed
// client-side. Fire-and-forget — any failure is swallowed and retried
// next launch. Zero behavioral impact if the network is unreachable.
//
// Privacy: the raw machine_id never leaves the device. See the design
// spec at docs/superpowers/specs/2026-05-01-device-id-analytics-design.md
// for the threat model.
import { app } from "electron";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { randomUUID, createHmac } from "node:crypto";
import { machineIdSync } from "node-machine-id";
import { ANALYTICS_SALT } from "./analytics-salt";
import { isSmokeTest } from "./smoke-probe";

// WHY: Moved to its own domain so Cloudflare's cache and rate limiter apply; the old workers.dev address still answers for older app versions.
const API_BASE = "https://api.youcoded.ai";
const ANALYTICS_FILE = path.join(os.homedir(), ".claude", "youcoded-analytics.json");

interface AnalyticsState {
  optIn: boolean;
  lastPingedDate: string;       // YYYY-MM-DD UTC, or "" when never pinged
  fallbackDeviceId?: string;    // present only if machine_id read failed
}

function defaultState(): AnalyticsState {
  return { optIn: true, lastPingedDate: "" };
}

function mapOs(platform: NodeJS.Platform): string {
  if (platform === "win32") return "win";
  if (platform === "darwin") return "mac";
  if (platform === "linux") return "linux";
  return "";
}

function todayUtc(): string {
  return new Date().toISOString().slice(0, 10);
}

function readState(): AnalyticsState {
  try {
    const raw = fs.readFileSync(ANALYTICS_FILE, "utf8");
    const parsed = JSON.parse(raw) as Partial<AnalyticsState> & {
      installId?: string;
      installReported?: boolean;
    };
    // Drop legacy installId / installReported silently — they're meaningless
    // post-cutover. The on-disk file gets rewritten on the next writeState().
    return {
      optIn: typeof parsed.optIn === "boolean" ? parsed.optIn : true,
      lastPingedDate: typeof parsed.lastPingedDate === "string" ? parsed.lastPingedDate : "",
      fallbackDeviceId: typeof parsed.fallbackDeviceId === "string" ? parsed.fallbackDeviceId : undefined,
    };
  } catch {
    return defaultState();
  }
}

function writeState(state: AnalyticsState): void {
  fs.mkdirSync(path.dirname(ANALYTICS_FILE), { recursive: true });
  fs.writeFileSync(ANALYTICS_FILE, JSON.stringify(state, null, 2));
}

export function getOptIn(): boolean {
  return readState().optIn;
}

export function setOptIn(value: boolean): void {
  const state = readState();
  state.optIn = value;
  writeState(state);
}

// Computes the device-id hash. Mutates state.fallbackDeviceId and persists
// the state file when the machine_id read fails.
export function deviceIdHash(state: AnalyticsState): string {
  let raw = "";
  try {
    raw = machineIdSync(true);  // `true` = return raw machine_id rather than the SHA-256 of it
  } catch {
    // swallowed — caught by the length check below
  }
  if (!raw || raw.length < 8) {
    if (!state.fallbackDeviceId) {
      state.fallbackDeviceId = randomUUID();
      writeState(state);
    }
    raw = `fallback:${state.fallbackDeviceId}`;
  }
  return createHmac("sha256", ANALYTICS_SALT)
    .update(`${raw}|${process.platform}`)
    .digest("hex");
}

async function postEvent(p: string, body: unknown): Promise<boolean> {
  try {
    const res = await fetch(`${API_BASE}${p}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    });
    return res.ok;
  } catch {
    return false;
  }
}

export async function runAnalyticsOnLaunch(): Promise<void> {
  // WHY: packaged CI smoke tests use ephemeral machines; counting them creates
  // phantom devices and retention cohorts, and must not mutate analytics state.
  if (isSmokeTest()) return;

  const state = readState();
  if (!state.optIn) return;

  const today = todayUtc();
  if (state.lastPingedDate === today) return;

  const hash = deviceIdHash(state);

  const ok = await postEvent("/app/heartbeat", {
    deviceIdHash: hash,
    appVersion: app.getVersion(),
    platform: "desktop" as const,
    os: mapOs(process.platform),
  });
  if (ok) {
    state.lastPingedDate = today;
    writeState(state);
  }
}

// WHY a cap: a sleeping computer pauses this countdown, so a single wait until
// midnight could run hours late after the lid opens. Each wake-up only looks at
// the clock; the network is used at most once per UTC day (runAnalyticsOnLaunch
// is gated on lastPingedDate).
const MAX_WAIT_MS = 3 * 60 * 60 * 1000;

// Milliseconds until the next UTC midnight (plus one second of margin so the
// check lands safely inside the new day), capped at MAX_WAIT_MS.
export function msUntilNextCheck(now: number = Date.now()): number {
  const d = new Date(now);
  const nextMidnight = Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate() + 1) + 1000;
  return Math.min(nextMidnight - now, MAX_WAIT_MS);
}

let dailyTimer: ReturnType<typeof setTimeout> | null = null;

// Sends today's heartbeat now (if not yet sent), then keeps sending one per
// UTC day for as long as the app stays open.
// WHY: the heartbeat used to fire only at launch, so an app left open for days
// was counted on its first day only.
export function startDailyHeartbeat(): void {
  if (dailyTimer) clearTimeout(dailyTimer);
  const tick = (): void => {
    // .catch first: a failed state-file write must neither surface as an
    // unhandled rejection nor stop the next day's check from being scheduled.
    void runAnalyticsOnLaunch().catch(() => {}).finally(() => {
      dailyTimer = setTimeout(tick, msUntilNextCheck());
      // Never keep the process alive just for analytics.
      dailyTimer.unref?.();
    });
  };
  tick();
}

File: app/src/main/kotlin/com/youcoded/app/analytics/AnalyticsService.kt
// Mirror of desktop/src/main/analytics-service.ts. Posts /app/heartbeat once
// per UTC day with a HMAC of (machine_id || platform), computed locally.
// Fire-and-forget — network failures do not throw and do not mutate state.
//
// Privacy: the raw machine_id never leaves the device. See the design
// spec at docs/superpowers/specs/2026-05-01-device-id-analytics-design.md
// for the threat model.
package com.youcoded.app.analytics

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.TimeZone
import java.util.UUID
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

data class AnalyticsState(
    val optIn: Boolean = true,
    val lastPingedDate: String = "",
    val fallbackDeviceId: String? = null,
)

class AnalyticsService(
    private val apiBase: String,
    private val homeDir: File,
    private val appVersion: String,
    private val http: OkHttpClient = OkHttpClient(),
    // Default reader fails loudly — production callers MUST inject a real reader
    // (typically a Settings.Secure.ANDROID_ID lookup; see SessionService.kt).
    private val machineIdReader: () -> String = ::defaultReaderUnimplemented,
) {
    private val stateFile get() = File(homeDir, ".claude/youcoded-analytics.json")

    fun runOnLaunch() {
        var state = readState()
        if (!state.optIn) return

        val today = todayUtc()
        if (state.lastPingedDate == today) return

        val (newState, hash) = computeDeviceIdHash(state)
        state = newState

        val payload = JSONObject().apply {
            put("deviceIdHash", hash)
            put("appVersion", appVersion)
            put("platform", "android")
            put("os", "")
        }
        if (postEvent("/app/heartbeat", payload)) {
            state = state.copy(lastPingedDate = today)
            writeState(state)
        }
    }

    fun getOptIn(): Boolean = readState().optIn

    fun setOptIn(value: Boolean) {
        val state = readState()
        writeState(state.copy(optIn = value))
    }

    // Internal helper for tests.
    fun debugReadState(): AnalyticsState = readState()

    private fun computeDeviceIdHash(state: AnalyticsState): Pair<AnalyticsState, String> {
        var current = state
        var raw = try { machineIdReader() } catch (_: Exception) { "" }
        if (raw.length < 8) {
            if (current.fallbackDeviceId == null) {
                current = current.copy(fallbackDeviceId = UUID.randomUUID().toString())
                writeState(current)
            }
            raw = "fallback:${current.fallbackDeviceId}"
        }
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(ANALYTICS_SALT.toByteArray(), "HmacSHA256"))
        val digest = mac.doFinal("$raw|android".toByteArray())
        val hash = digest.joinToString("") { "%02x".format(it) }
        return current to hash
    }

    private fun readState(): AnalyticsState {
        if (!stateFile.exists()) return AnalyticsState()
        return try {
            val json = JSONObject(stateFile.readText())
            // Drop legacy installId / installReported silently; they're meaningless
            // post-cutover and the file gets rewritten on the next writeState().
            AnalyticsState(
                optIn = json.optBoolean("optIn", true),
                lastPingedDate = json.optString("lastPingedDate", ""),
                fallbackDeviceId = if (json.has("fallbackDeviceId") && !json.isNull("fallbackDeviceId"))
                    json.getString("fallbackDeviceId") else null,
            )
        } catch (_: Exception) {
            AnalyticsState()
        }
    }

    private fun writeState(state: AnalyticsState) {
        stateFile.parentFile?.mkdirs()
        val json = JSONObject().apply {
            put("optIn", state.optIn)
            put("lastPingedDate", state.lastPingedDate)
            if (state.fallbackDeviceId != null) put("fallbackDeviceId", state.fallbackDeviceId)
        }
        stateFile.writeText(json.toString(2))
    }

    private fun postEvent(path: String, body: JSONObject): Boolean {
        return try {
            val req = Request.Builder()
                .url(apiBase + path)
                .post(body.toString().toRequestBody("application/json".toMediaType()))
                .build()
            http.newCall(req).execute().use { it.isSuccessful }
        } catch (_: Exception) {
            false
        }
    }

    companion object {
        // WHY a cap: a sleeping phone pauses the countdown, so one wait until
        // midnight could run hours late. Each wake-up only looks at the clock;
        // runOnLaunch() sends at most once per UTC day. Mirrors desktop's
        // msUntilNextCheck in analytics-service.ts.
        private const val MAX_WAIT_MS = 3L * 60 * 60 * 1000
        private const val DAY_MS = 24L * 60 * 60 * 1000

        fun todayUtc(): String {
            val fmt = SimpleDateFormat("yyyy-MM-dd").apply { timeZone = TimeZone.getTimeZone("UTC") }
            return fmt.format(Date())
        }

        // Milliseconds until one second past the next UTC midnight, capped at 3 hours.
        // Epoch milliseconds count UTC days exactly, so no calendar is needed.
        fun msUntilNextCheck(nowMs: Long): Long {
            val nextMidnight = (nowMs / DAY_MS + 1) * DAY_MS + 1000
            return minOf(nextMidnight - nowMs, MAX_WAIT_MS)
        }
    }
}

// Sentinel for the default constructor argument. Production callers MUST
// inject a real reader at construction time (see SessionService.kt where
// AnalyticsService is instantiated with a Settings.Secure.ANDROID_ID lambda).
private fun defaultReaderUnimplemented(): String {
    throw IllegalStateException(
        "AnalyticsService.machineIdReader was not injected. " +
        "Pass a Settings.Secure.ANDROID_ID lambda from your Android Context."
    )
}
